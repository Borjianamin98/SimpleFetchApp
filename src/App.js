import React, {Component} from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSpinner } from '@fortawesome/free-solid-svg-icons'
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '100';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
    _isMounted = false;

    constructor(params) {
        super(params);
        this.state = {
            results: null,
            searchKey: '',
            searchTerm: DEFAULT_QUERY,
            error: null,
            isLoading: false,
        };

        this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
        this.setSearchTopStories = this.setSearchTopStories.bind(this);
        this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
        this.onDismiss = this.onDismiss.bind(this);
    }

    needsToSearchTopStories(searchTerm) {
        return !this.state.results[searchTerm];
    }

    setSearchTopStories(result) {
        const {hits, page} = result;
        const {searchKey, results} = this.state;
        const oldHits = results && results[searchKey]
            ? results[searchKey].hits
            : [];
        const updatedHits = [
            ...oldHits,
            ...hits
        ];
        this.setState({
            results: {
                ...results,
                [searchKey]: {
                    hits: updatedHits,
                    page
                }
            },
            isLoading: false
        });
    }

    fetchSearchTopStories(searchTerm, page = 0) {
        this.setState({ isLoading: true });
        axios.get(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${
            page}&${PARAM_HPP}${DEFAULT_HPP}`)
            .then(result => this.setSearchTopStories(result.data))
            .catch(error => this._isMounted && this.setState({error}));
    }

    onSearchChange(event) {
        this.setState({searchTerm: event.target.value});
    }

    onSearchSubmit(event) {
        const {searchTerm} = this.state;
        this.setState({searchKey: searchTerm});

        if (this.needsToSearchTopStories(searchTerm)) {
            this.fetchSearchTopStories(searchTerm);
        }

        event.preventDefault();
    }

    onDismiss(id) {
        const {searchKey, results} = this.state;
        const {hits, page} = results[searchKey];

        const isNotId = item => item.objectID !== id;
        const updatedHits = hits.filter(isNotId);

        this.setState({
            results: {
                ...results,
                [searchKey]: {hits: updatedHits, page}
            }
        });
    }

    componentDidMount() {
        this._isMounted = true;

        const {searchTerm} = this.state;
        this.setState({searchKey: searchTerm});
        this.fetchSearchTopStories(searchTerm);
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    render() {
        const {
            searchTerm,
            searchKey,
            results,
            error,
            isLoading
        } = this.state;

        const page = (
            results &&
            results[searchKey] &&
            results[searchKey].page
        ) || 0;

        const list = (
            results &&
            results[searchKey] &&
            results[searchKey].hits
        ) || [];

        return (
            <div className="page">
                <div className="interactions">
                    <Search
                        value={searchTerm}
                        onChange={this.onSearchChange}
                        onSubmit={this.onSearchSubmit}
                    >
                        Search
                    </Search>
                </div>
                {error
                    ? <div className="interactions">
                        <p>Something went wrong.</p>
                    </div>
                    : <Table
                        list={list}
                        onDismiss={this.onDismiss}
                    />
                }
                <div className="interactions">
                    { isLoading
                        ? <Loading />
                        : <Button
                            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
                        >
                            More
                        </Button>
                    }
                </div>
            </div>
        );
    }
}

const Loading = () =>
    <FontAwesomeIcon icon={faSpinner} spin size="3x" />

class Search extends Component {
    componentDidMount() {
        if (this.input) {
            this.input.focus();
        }
    }

    render() {
        const {value, onChange, onSubmit, children} = this.props;
        return (
            <form onSubmit={onSubmit}>
                <input
                    type="text"
                    value={value}
                    onChange={onChange}
                    ref={instance => this.input = instance}
                />
                <button type="submit">
                    {children}
                </button>
            </form>
        );
    }
}

Search.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
    children: PropTypes.node.isRequired,
}

const largeColumn = {
    width: '40%',
};
const midColumn = {
    width: '30%',
};
const smallColumn = {
    width: '10%',
};

const Table = ({list, onDismiss}) =>
    <div className="table">
        {list.map(item =>
            <div key={item.objectID} className="table-row">
                <span style={largeColumn}>
                    <a href={item.url}>{item.title}</a>
                </span>
                <span style={midColumn}>{item.author}</span>
                <span style={smallColumn}>{item.num_comments}</span>
                <span style={smallColumn}>{item.points}</span>
                <span style={smallColumn}>
                    <Button
                        onClick={() => onDismiss(item.objectID)}
                        className="button-inline">
                        Dismiss
                    </Button>
                </span>
            </div>
        )}
    </div>;

Table.propTypes = {
    list: PropTypes.arrayOf(PropTypes.shape({
        objectID: PropTypes.string.isRequired,
        title: PropTypes.string,
        url: PropTypes.string,
        author: PropTypes.string,
        num_comments: PropTypes.number,
        points: PropTypes.number,
    })).isRequired,
    onDismiss : PropTypes.func.isRequired
}

const Button = ({onClick, className, children}) =>
    <button
        onClick={onClick}
        className={className}
        type="button"
    >
        {children}
    </button>;

Button.defaultProps = {
    className: '',
};

Button.propTypes = {
    onClick: PropTypes.func.isRequired,
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
};

export default App;
export {
    Button,
    Search,
    Table,
};