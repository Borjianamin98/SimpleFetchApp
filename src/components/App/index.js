import React from 'react';
import axios from 'axios';
import Table from "../Table";
import Search from "../Search";
import {ButtonWithLoading} from "../Button";
import './index.css';
import {
    DEFAULT_QUERY,
    DEFAULT_HPP,
    PATH_BASE,
    PATH_SEARCH,
    PARAM_SEARCH,
    PARAM_PAGE,
    PARAM_HPP
} from '../../constants'

const useMountEffect = (func) => React.useEffect(func, [])
const updateSearchTopStoriesState = (hits, page) => (prevTopResult) => {
    const {results, searchKey} = prevTopResult;
    const oldHits = results && results[searchKey]
        ? results[searchKey].hits
        : [];
    const updatedHits = [
        ...oldHits,
        ...hits
    ];
    return {
        results: {
            ...results,
            [searchKey]: {hits: updatedHits, page}
        },
        searchKey
    };
};

const App = () => {
    const isMounted = React.useRef(false);
    const [topResult, setTopResult] = React.useState({
        results: null,
        searchKey: DEFAULT_QUERY,
    });
    const [searchTerm, setSearchTerm] = React.useState(DEFAULT_QUERY);
    const [error, setError] = React.useState(null);
    const [isLoading, setIsLoading] = React.useState(false);

    function needsToSearchTopStories(searchTerm) {
        return !topResult.results[searchTerm];
    }

    function setSearchTopStories(result) {
        const {hits, page} = result;
        setTopResult(updateSearchTopStoriesState(hits, page));
        setIsLoading(false);
    }

    function fetchSearchTopStories(searchTerm, page = 0) {
        setIsLoading(true);
        axios.get(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${
            page}&${PARAM_HPP}${DEFAULT_HPP}`)
            .then(result => isMounted.current && setSearchTopStories(result.data))
            .catch(error => isMounted.current && setError(error));
    }

    function onSearchChange(event) {
        setSearchTerm(event.target.value)
    }

    function onSearchSubmit(event) {
        setTopResult({
            ...topResult,
            searchKey: searchTerm
        })

        if (needsToSearchTopStories(searchTerm)) {
            fetchSearchTopStories(searchTerm);
        }

        event.preventDefault();
    }

    function onDismiss(id) {
        const {hits, page} = topResult.results[topResult.searchKey];
        const updatedHits = hits.filter(item => item.objectID !== id);
        const newTopResult = {...topResult};
        newTopResult.results[[topResult.searchKey]] = {hits: updatedHits, page};
        setTopResult(newTopResult)
    }

    useMountEffect(() => {
        isMounted.current = true;
        fetchSearchTopStories(DEFAULT_QUERY);

        return () => {
            isMounted.current = false;
        };
    })

    const {results, searchKey} = topResult;
    const page = (
        results &&
        results[searchKey] &&
        results[searchKey].page
    ) || 0;

    const list = (
        results &&
        results[searchKey] &&
        results[searchKey].hits
    ) || [];

    return (
        <div className="page">
            <div className="interactions">
                <Search
                    value={searchTerm}
                    onChange={onSearchChange}
                    onSubmit={onSearchSubmit}
                >
                    Search
                </Search>
            </div>
            {error
                ? <div className="interactions">
                    <p>Something went wrong.</p>
                </div>
                : <Table
                    list={list}
                    onDismiss={onDismiss}
                />
            }
            <div className="interactions">
                <ButtonWithLoading
                    isLoading={isLoading}
                    onClick={() => fetchSearchTopStories(searchKey, page + 1)}
                >
                    More
                </ButtonWithLoading>
            </div>
        </div>
    );
}

export default App;