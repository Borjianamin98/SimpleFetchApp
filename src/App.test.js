import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { Search, Button, Table } from './App';

describe('App', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
        <Search>Search</Search>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give Me More</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  test('has a valid snapshot', () => {
    const component = renderer.create(
        <Button>Give Me More</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
  };
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table { ...props } />, div);
  });
  test('has a valid snapshot', () => {
    const component = renderer.create(
        <Table { ...props } />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});