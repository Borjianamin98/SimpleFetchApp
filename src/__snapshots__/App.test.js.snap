// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-3x "
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give Me More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value="defaultValue"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
